### DISCUSSION TITLE
Anybody using tmux + textual?

### DISCUSSION THREAD

**Discussion by lauralindzey** [Textualize/textual#4003] [Q&A]
I'd like to be able to run our TUI in tmux, but am having two (non-show-stopping, but still annoying) issues. So I figured I'd ask whether anybody else has workarounds/fixes for these.

~~~
$ tmux -V
> tmux 3.0a
~~~

1. I've bound the escape key to de-focus Input fields. However, when running in tmux, it takes > 1 second for this callback to happen, though all others are snappy. In the default ubuntu terminal, the `escape` binding's effect is instant.

Relevant code:
~~~python
BINDINGS = [
    ...
    Binding("escape", "unfocus", "Unfocus text input", show=False),
]
def action_unfocus(self) -> None:
    """
    Unfocus text entry boxes.
    I can't figure out how to unfocus them without selecting another
    widget for focus, so just pick a container to set focus, from which the
    rest of the keybindings will work.
    """
    sc = self.query_one(ScrollableContainer)
    sc.focus()
~~~

2. It seems that tmux doesn't support all of the colors used by Textual. Is there a configuration (either for textual, limiting the colors used, or tmux, supporting truecolor) that works better?

Here's an example showing the issue:

<img width="716" alt="Screenshot 2024-01-10 at 17 24 34" src="https://github.com/Textualize/textual/assets/65185744/d11f6c71-6668-4016-a69c-a195aadbbee7">

From left, we have:
* Running in default ubuntu terminal, start disabled, others enabled
* Running in default ubuntu terminal, start enabled, others disabled
* Running in tmux, start disabled, others enabled
* Running in tmux, start enabled, others disabled

~~~python
#! /usr/bin/env python3

from textual.app import App, ComposeResult
from textual.widgets import Button


class ButtonDisplayApp(App):
    def compose(self) -> ComposeResult:
        yield Button("Start", id="start", variant="primary")
        yield Button("Success", id="success", variant="success", disabled=True)
        yield Button("Cancel", id="cancel", variant="warning", disabled=True)
        yield Button("Fail", id="fail", variant="error", disabled=True)

    def on_button_pressed(self, event: Button.Pressed) -> None:
        started = [event.button.id](http://event.button.id/) == "start"
        for button in self.query(Button):
            if [button.id](http://button.id/) == "start":
                button.disabled = started
            else:
                button.disabled = not started


if __name__ == "__main__":
    app = ButtonDisplayApp()
    app.run()
~~~

**Comment by 1dancook** [comment]
Hi! I'm running TMUX 3.3a and I've been using one textual app I made for many hours every day.

The short version of my response: what you are experiencing has nothing to do with textual. 

More information:

- I am on macOS
- Terminal Emulator is Kitty

### Problem 1 - `ESCAPE` time

In my gradebook app (sorry, this isn't public) I have some different forms that require this exact behavior (escape from an input and set the focus somewhere else). It works as expected without delay.

You might try a different terminal emulator to see if the emulator is the source of this issue or not.

### Problem 2 - Colors

I have a few settings in TMUX for more colors, so I can't reproduce the kind of color issue you showed in your screenshot. 

I have some settings in my ~/.tmux.conf specifically to enable more colors.

My suggestion here is to check online for the settings that you might require for your setup and apply them to your tmux config.
